File: kotlin/protocol/src/main/kotlin/org/hpb/protocol/Validation.kt
package org.hpb.protocol

import kotlinx.serialization.json.jsonObject
import org.hpb.engine.hex
import org.hpb.engine.nostr.Events
import org.hpb.engine.nostr.NostrEvent
import org.hpb.engine.sha256
import org.hpb.protocol.Pj.a
import org.hpb.protocol.Pj.b
import org.hpb.protocol.Pj.s

/** One revealed, validated answer in the final results set. */
data class ResultRow(
    val taskKey: String,
    val worker: String,
    val answer: String,
    val accepted: Boolean,
)

/**
 * The launcher's reveal: the full validated submission set, INLINE. Anyone
 * recomputes validation and payout correctness from this event; its sha256
 * is the results hash committed in the on-chain PAYOUT record. Non-reveal is
 * a provable reputational failure (workers hold their signed submissions).
 */
data class EscrowResults(val escrowId: String, val rows: List<ResultRow>) {
    fun resultsJson(): String = Pj.obj(
        "v" to Pj.num(ProtocolKinds.VERSION),
        "rows" to Pj.arr(
            rows.map {
                Pj.obj(
                    "task_key" to Pj.str(it.taskKey),
                    "worker" to Pj.str(it.worker),
                    "answer" to Pj.str(it.answer),
                    "accepted" to Pj.bool(it.accepted),
                )
            },
        ),
    ).toString()

    fun resultsHash(): ByteArray = sha256(resultsJson().toByteArray())
}

object Validations {
    fun toEvent(privkey: ByteArray, results: EscrowResults, createdAt: Long): NostrEvent =
        Events.sign(
            privkey, ProtocolKinds.VALIDATION,
            tags = listOf(listOf("x", results.escrowId)),
            content = results.resultsJson(),
            createdAt = createdAt,
        )

    fun fromEvent(event: NostrEvent): EscrowResults {
        require(event.kind == ProtocolKinds.VALIDATION) { "not a validation" }
        val content = Pj.parse(event.content)
        return EscrowResults(
            escrowId = requireNotNull(event.tagValue("x")),
            rows = content.a("rows").map {
                val row = it.jsonObject
                ResultRow(row.s("task_key"), row.s("worker"), row.s("answer"), row.b("accepted"))
            },
        )
    }

    fun resultsHashHex(results: EscrowResults): String = results.resultsHash().hex()
}

File: ios/HpbCore/Sources/HpbCore/Validation.swift
import Foundation

/// One revealed, validated answer in the final results set.
public struct ResultRow: Equatable {
    public let taskKey: String
    public let worker: String
    public let answer: String
    public let accepted: Bool

    public init(taskKey: String, worker: String, answer: String, accepted: Bool) {
        self.taskKey = taskKey
        self.worker = worker
        self.answer = answer
        self.accepted = accepted
    }
}

/// The launcher's reveal: the full validated submission set, INLINE; its
/// sha256 is the results hash committed in the on-chain PAYOUT record.
public struct EscrowResults {
    public let escrowId: String
    public let rows: [ResultRow]

    public init(escrowId: String, rows: [ResultRow]) {
        self.escrowId = escrowId
        self.rows = rows
    }

    public func resultsJson() -> String {
        Cj.write(Cj.obj([
            ("v", Cj.num(ProtocolKinds.version)),
            ("rows", .arr(rows.map {
                Cj.obj([
                    ("task_key", .str($0.taskKey)),
                    ("worker", .str($0.worker)),
                    ("answer", .str($0.answer)),
                    ("accepted", .bool($0.accepted)),
                ])
            })),
        ]))
    }

    public func resultsHash() -> [UInt8] {
        sha256(resultsJson())
    }
}

public enum Validations {
    public static func toEvent(_ privkey: [UInt8], _ results: EscrowResults, createdAt: Int64) throws -> NostrEvent {
        try Events.sign(
            privkey, kind: ProtocolKinds.validation,
            tags: [["x", results.escrowId]],
            content: results.resultsJson(),
            createdAt: createdAt
        )
    }

    public static func fromEvent(_ event: NostrEvent) throws -> EscrowResults {
        guard event.kind == ProtocolKinds.validation else { throw HpbError.invalid("not a validation") }
        let content = try Cj.parse(event.content)
        guard let escrowId = event.tagValue("x") else { throw HpbError.invalid("validation missing x tag") }
        return EscrowResults(
            escrowId: escrowId,
            rows: try content.a("rows").map {
                ResultRow(
                    taskKey: try $0.s("task_key"),
                    worker: try $0.s("worker"),
                    answer: try $0.s("answer"),
                    accepted: try $0.b("accepted")
                )
            }
        )
    }
}
